RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h

```diff
@@ -264,6 +305,8 @@ class PixelCPEBase : public PixelClusterParameterEstimator {
 
   static constexpr float clusterSplitMaxError_ = 7777.7f;
 
+  mutable PixelCPEWarningSummary warnings_;
```

Comment:
just declaring something mutable doesn't make it thread-safe (in fact, the opposite). if this object will potentially be modified by multiple threads, you need to ensure there are no data races, ideally by using atomic operations.

Reply:
Yes I didn't really think about thread safety when doing this. I made it mutable so that we can add messages to the warnings from the const local_position function. I think you are right that there could be data races but given this is just a warning summary printout I think the only races could be the warning getting printed twice. Still if this is an issue I can try to use atomic operations in the warning object but I am not that familiar with them. Is there a way to atomically add/read things from a map? I am also open to other implementation suggestions

Reply:
Whether or not this is thread-safe depends on how this object is used. The `WarningSummary` from the unpacker is only used in a stream producer, so each instance will only be accessed by one thread. I'm not familiar enough with the details of the pixel reco code to know if that is also the case here.

In general, atomic operations on complicated objects should be done using `compare_exchange_strong`. The basic idea is a compare-and-swap loop: make the desired change to a copy of the original object, then try to make an atomic update of the original object from the copy. There are examples in CMSSW and elsewhere. (Of course, all the copying and comparing can decrease performance, so it should be assessed whether this is really necessary in this case.)

Reply:
> I'm not familiar enough with the details of the pixel reco code to know if that is also the case here.

Classes derived from `PicelCPEBase` are EventSetup products, so thread safety and efficiency are of uttermost importance.

Reply:
> use tbb::concurrent_unordered_map (for an atomic "insert-or-find") with std::atomic<std::size_t> for the counter (even though it would require wrapping the atomic to custom type that allows copying).

Ok thanks for the suggestion. I just pushed a new commit based on this which ran fine for me, let me know if I did something incorrectly

Reply:
For an exact count of messages the separate `find()` and `emplace()` are not sufficient, because another thread could insert the same key into the map between the `find()` and `emplace()` of the current thread, in which case the `emplace()` of the current thread gets ignored. If an exact is not important and an approximate suffices, the current pattern is ok from thread-safety point of view.

For an exact count the fact that `emplace()` returns an iterator to the existing element in case they key is found can be used to have a single atomic "find-or-emplace" operation along
```cpp
auto iteratorInsertedPair = map.emplace("foo"s, 0);
++(iteratorInsertedPair.first->second);
```

Written all that, I'm a bit concerned of collecting bunch of data into an ESProduct (even if the "visible state" of the ESProduct does not really change, at least what concerns physics). How big could this map grow in the works case? (appears to depend how many different `theClusterParam.ierr` values there can be)

Reply:
Thanks for the explanation. I think that an exact count of warning summary is not needed and an approximate completely does the job. This will only be used to check if there are large numbers of a particular type of error, indicating something is wrong. 

Looking through: https://github.com/cms-sw/cmssw/blob/master/RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco.cc I count 9 different error codes so in the absolute worse case the map could be size 9. I will also add that realistically I have never seen more than one or two types of errors in thrown even when there is some issue with the templates so really I would expect the map to say very small nearly always.